File: code-checker/src/main/java/org/example/code/checker/checker/markdown/task/checker/structrue/front/matter/FrontMatterChecker.java
package org.example.code.checker.checker.markdown.task.checker.structrue.front.matter;

import org.example.code.checker.checker.CheckError;
import org.example.flow.engine.node.TaskData;
import org.example.flow.engine.node.TaskDataUtils;
import org.example.flow.engine.node.TaskNode;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Check FrontMatter block at the beginning of markdown document.
 * <p>
 * Rules:
 * <ul>
 *     <li>Document must contain a YAML FrontMatter block delimited by {@code ---}.</li>
 *     <li>FrontMatter keys must match the attributes defined in
 *     {@code huaweicloud_cdn_ip_information.md}:
 *     {@code subcategory}, {@code layout}, {@code page_title}, {@code description}.</li>
 * </ul>
 * <p>
 * On success this task returns {@code TaskData<FrontMatter>} with parsed values.
 * On failure it returns {@code TaskData<CheckError>} describing the problem.
 */
public class FrontMatterChecker extends TaskNode {

    private static final Set<String> REQUIRED_KEYS = new HashSet<>(
        Arrays.asList("subcategory", "layout", "page_title", "description")
    );

    @Override
    public TaskData<?> task(Map<String, TaskData<?>> input) {
        if (input == null || input.isEmpty()) {
            throw new IllegalArgumentException("input is null or empty");
        }

        String rawCode = TaskDataUtils.getPayload(input, "rawCode", String.class);
        String fileId = TaskDataUtils.getPayload(input, "fileId", String.class);

        Set<String> frontMatterKeys = extractFrontMatterKeys(rawCode);
        if (frontMatterKeys.isEmpty()) {
            CheckError error = new CheckError(
                "FrontMatter.Missing",
                "FrontMatter block is missing in markdown document",
                CheckError.Severity.ERROR,
                fileId,
                null,
                null,
                "FrontMatter"
            );
            return new TaskData<>(
                CheckError.class,
                FrontMatterChecker.class.getSimpleName(),
                System.currentTimeMillis(),
                error
            );
        }

        Set<String> missing = new HashSet<>(REQUIRED_KEYS);
        missing.removeAll(frontMatterKeys);
        if (!missing.isEmpty()) {
            CheckError error = new CheckError(
                "FrontMatter.MissingKeys",
                "FrontMatter keys are not consistent with template. Missing keys: " + missing,
                CheckError.Severity.ERROR,
                fileId,
                null,
                null,
                "FrontMatter"
            );
            return new TaskData<>(
                CheckError.class,
                FrontMatterChecker.class.getSimpleName(),
                System.currentTimeMillis(),
                error
            );
        }

        FrontMatter frontMatter = parseFrontMatter(rawCode);

        return new TaskData<>(
            FrontMatter.class,
            FrontMatterChecker.class.getSimpleName(),
            System.currentTimeMillis(),
            frontMatter
        );
    }

    /**
     * Extract FrontMatter keys from the given markdown text.
     * The FrontMatter is expected to be a YAML block at the beginning:
     *
     * <pre>
     * ---
     * key1: value
     * key2: "value"
     * ...
     * ---
     * </pre>
     *
     * @param markdown full markdown content
     * @return set of key names found in FrontMatter, or empty set if block not found
     */
    private Set<String> extractFrontMatterKeys(String markdown) {
        Set<String> keys = new HashSet<>();
        if (markdown == null || markdown.isEmpty()) {
            return keys;
        }

        String[] lines = markdown.split("\\r?\\n");
        if (lines.length == 0) {
            return keys;
        }

        int idx = 0;
        // FrontMatter must start at the very beginning with '---'
        if (!"---".equals(lines[idx].trim())) {
            return keys;
        }
        idx++;

        for (; idx < lines.length; idx++) {
            String line = lines[idx].trim();
            if ("---".equals(line)) {
                // End of FrontMatter block
                break;
            }
            if (line.isEmpty() || line.startsWith("#")) {
                continue;
            }
            int colon = line.indexOf(':');
            if (colon <= 0) {
                continue;
            }
            String key = line.substring(0, colon).trim();
            if (!key.isEmpty()) {
                keys.add(key);
            }
        }

        return keys;
    }

    /**
     * Parse FrontMatter values from markdown and build {@link FrontMatter}.
     * Only the fields we care about (subcategory, layout, page_title, description)
     * are extracted.
     */
    private FrontMatter parseFrontMatter(String markdown) {
        if (markdown == null || markdown.isEmpty()) {
            return null;
        }

        String[] lines = markdown.split("\\r?\\n");
        if (lines.length == 0) {
            return null;
        }

        int idx = 0;
        if (!"---".equals(lines[idx].trim())) {
            return null;
        }
        idx++;

        String subcategory = null;
        String layout = null;
        String pageTitle = null;
        String description = null;

        boolean inDescriptionBlock = false;
        StringBuilder descBuilder = null;

        for (; idx < lines.length; idx++) {
            String rawLine = lines[idx];
            String line = rawLine.trim();

            if ("---".equals(line)) {
                break;
            }

            if (inDescriptionBlock) {
                if (line.startsWith("---")) {
                    break;
                }
                if (rawLine.startsWith(" ") || rawLine.startsWith("\t")) {
                    if (descBuilder.length() > 0) {
                        descBuilder.append('\n');
                    }
                    descBuilder.append(line);
                    continue;
                } else {
                    break;
                }
            }

            if (line.isEmpty() || line.startsWith("#")) {
                continue;
            }

            int colon = line.indexOf(':');
            if (colon <= 0) {
                continue;
            }
            String key = line.substring(0, colon).trim();
            String valuePart = line.substring(colon + 1).trim();

            if ("subcategory".equals(key)) {
                subcategory = stripYamlQuotes(valuePart);
            } else if ("layout".equals(key)) {
                layout = stripYamlQuotes(valuePart);
            } else if ("page_title".equals(key)) {
                pageTitle = stripYamlQuotes(valuePart);
            } else if ("description".equals(key)) {
                if (valuePart.startsWith("|")) {
                    inDescriptionBlock = true;
                    descBuilder = new StringBuilder();
                } else {
                    description = stripYamlQuotes(valuePart);
                }
            }
        }

        if (description == null && descBuilder != null) {
            description = descBuilder.toString().trim();
        }

        return new FrontMatter(subcategory, layout, pageTitle, description);
    }

    private String stripYamlQuotes(String value) {
        if (value == null || value.isEmpty()) {
            return value;
        }
        if ((value.startsWith("\"") && value.endsWith("\""))
            || (value.startsWith("'") && value.endsWith("'"))) {
            return value.substring(1, value.length() - 1);
        }
        return value;
    }
}

File: code-checker/src/main/kotlin/org/example/code/checker/checker/markdown/task/checker/structrue/front/matter/FrontMatterChecker.kt
package org.example.code.checker.checker.markdown.task.checker.structrue.front.matter

import org.example.code.checker.checker.common.CheckError
import org.example.code.checker.checker.markdown.domain.MdDomain
import org.example.code.checker.checker.markdown.domain.standard.block.Heading
import org.example.code.checker.checker.utils.TreeNode
import org.example.flow.engine.node.TaskData
import org.example.flow.engine.node.TaskDataUtils
import org.example.flow.engine.node.TaskNode

/**
 * Front matter checking and consuming rule.
 *
 * Responsibilities:
 * - Analyse the YAML-like front matter block at the top of the markdown file.
 * - Validate that the description field follows the expected pattern.
 * - Remove front matter blocks from the working document tree so that
 *   downstream rules (e.g. TitleChecker) only see the body.
 *
 * Input:
 * - rawCode:          TaskData<String>             – entire markdown source.
 * - originalDocument: TaskData<TreeNode<MdDomain>> – original document tree (read-only).
 * - workingDocument:  TaskData<TreeNode<MdDomain>> – working document tree to be mutated.
 * - fileId:           TaskData<String>             – file identifier or path.
 *
 * Output:
 * - TaskData<TreeNode<MdDomain>> – updated working document tree.
 * - TaskData<List<CheckError>>   – front-matter related errors.
 */
class FrontMatterChecker : TaskNode() {

    override fun task(input: Map<String, TaskData<*>>?): List<TaskData<*>> {
        if (input == null || input.isEmpty()) {
            throw IllegalArgumentException("input is null or empty")
        }

        val rawCode: String =
            TaskDataUtils.getPayload(input, "rawCode", String::class.java)

        @Suppress("UNCHECKED_CAST")
        val originalDocument: TreeNode<MdDomain> =
            TaskDataUtils.getPayload(input, "originalDocument", TreeNode::class.java) as TreeNode<MdDomain>

        @Suppress("UNCHECKED_CAST")
        val workingDocument: TreeNode<MdDomain> =
            TaskDataUtils.getPayload(input, "workingDocument", TreeNode::class.java) as TreeNode<MdDomain>

        val fileId: String =
            TaskDataUtils.getPayload(input, "fileId", String::class.java)

        val errors = mutableListOf<CheckError>()

        // 1. Text-level validation for front matter (mainly description field).
        errors += checkFrontMatterDescription(fileId, rawCode)

        // 2. Structural consumption: remove front-matter blocks from working tree.
        consumeFrontMatter(workingDocument)

        val now = System.currentTimeMillis()
        val outputs = mutableListOf<TaskData<*>>()

        // Updated working document.
        outputs += TaskData(
            TreeNode::class.java,
            "${FrontMatterChecker::class.java.simpleName}.workingDocument",
            now,
            workingDocument
        )

        // Front-matter check results (can be empty).
        outputs += TaskData(
            List::class.java,
            "${FrontMatterChecker::class.java.simpleName}.errors",
            now,
            errors.toList()
        )

        return outputs
    }

    /**
     * Validate the description block in YAML-like front matter.
     *
     * Expectation (simplified based on doc_rule.md):
     * - There is a front matter block at the very top, delimited by lines "---".
     * - Inside the block there is a description field with the following shape:
     *
     *   description: |-
     *     Use this (data source|resource) to ... within HuaweiCloud.
     */
    private fun checkFrontMatterDescription(
        fileId: String,
        source: String
    ): List<CheckError> {
        val errors = mutableListOf<CheckError>()
        if (source.isEmpty()) {
            return errors
        }

        val lines = source.split('\n')
        if (lines.isEmpty()) {
            return errors
        }

        // Locate front matter block: between first and second '---' at top of file.
        if (!lines[0].trim().startsWith("---")) {
            return errors // no front matter; nothing to check.
        }

        var secondDelimiterIndex = -1
        for (i in 1 until lines.size) {
            if (lines[i].trim().startsWith("---")) {
                secondDelimiterIndex = i
                break
            }
        }
        if (secondDelimiterIndex <= 0) {
            return errors // malformed or incomplete front matter; skip strict checks.
        }

        val frontMatterLines = lines.subList(1, secondDelimiterIndex)

        // Find the 'description: |-' line and the first non-empty line after it.
        var descriptionLineIndex = -1
        for (i in frontMatterLines.indices) {
            if (frontMatterLines[i].trim().startsWith("description:")) {
                descriptionLineIndex = i
                break
            }
        }
        if (descriptionLineIndex < 0) {
            errors += CheckError.builder()
                .ruleId("FrontMatter.MissingDescription")
                .message("Front matter description field is missing")
                .severity(CheckError.Severity.ERROR)
                .fileId(fileId)
                .range(null)
                .nodeId(null)
                .nodeType("FrontMatter")
                .build()
            return errors
        }

        val descHeader = frontMatterLines[descriptionLineIndex].trim()
        if (!descHeader.matches(Regex("""^description:\s*\|-\s*$"""))) {
            errors += CheckError.builder()
                .ruleId("FrontMatter.DescriptionFormat")
                .message("Front matter description must use block scalar syntax: 'description: |-'")
                .severity(CheckError.Severity.ERROR)
                .fileId(fileId)
                .range(null)
                .nodeId(null)
                .nodeType("FrontMatter")
                .build()
            return errors
        }

        var descriptionText: String? = null
        for (i in (descriptionLineIndex + 1) until frontMatterLines.size) {
            val line = frontMatterLines[i]
            if (line.isBlank()) continue
            // Remove leading indentation (commonly two spaces).
            descriptionText = line.trim()
            break
        }

        if (descriptionText == null || descriptionText.isEmpty()) {
            errors += CheckError.builder()
                .ruleId("FrontMatter.EmptyDescription")
                .message("Front matter description block is empty")
                .severity(CheckError.Severity.ERROR)
                .fileId(fileId)
                .range(null)
                .nodeId(null)
                .nodeType("FrontMatter")
                .build()
            return errors
        }

        val normalized = descriptionText.trim()

        // Very lightweight semantic pattern: "Use this (data source|resource) to ... within HuaweiCloud."
        val pattern = Regex(
            pattern = """^Use this\s+(data source|resource)\s+to\s+.+\s+within\s+HuaweiCloud\.\s*$""",
            option = RegexOption.IGNORE_CASE
        )

        if (!pattern.matches(normalized)) {
            errors += CheckError.builder()
                .ruleId("FrontMatter.DescriptionSemantic")
                .message("Front matter description should be 'Use this (data source|resource) to ... within HuaweiCloud.'")
                .severity(CheckError.Severity.WARNING)
                .fileId(fileId)
                .range(null)
                .nodeId(null)
                .nodeType("FrontMatter")
                .build()
        }

        return errors
    }

    /**
     * Remove the leading block-level nodes that represent front matter
     * from the working document tree.
     *
     * Current heuristic:
     * - Front matter is any sequence of top-level blocks before the first
     *   level-1 heading (Heading level == 1).
     * - These nodes are removed from the root's children list.
     */
    private fun consumeFrontMatter(workingDocument: TreeNode<MdDomain>) {
        val children = workingDocument.children ?: return
        if (children.isEmpty()) return

        var firstHeadingIndex = -1
        for (i in children.indices) {
            val data = children[i].data
            if (data is Heading && data.level == 1) {
                firstHeadingIndex = i
                break
            }
        }

        if (firstHeadingIndex <= 0) {
            // No heading or heading already at the top; nothing to consume.
            return
        }

        val toRemove = children.subList(0, firstHeadingIndex).toList()
        for (child in toRemove) {
            workingDocument.removeChild(child)
        }
    }
}


